# SUMMER ANALYTICS 2022

## WEEK-4 ASSIGNMENT

In [1]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import matplotlib.pyplot as plt # data visualization
import seaborn as sns # statistical data visualization
from sklearn.preprocessing import LabelEncoder
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

**Let's explore the dataset shall we?**

The Car Evaluation Database contains examples with the structural information removed, i.e., directly relates CAR to the six input attributes: buying, maintenance cost, doors, persons, lug_boot, safety.
In this notebook we will go through an in depth analysis of sound and how we can classify and ultimately understand it.


In [3]:
# Read the csv file as data
df = pd.read_csv("car_evaluation.csv")
df.head(10)

vhigh vhigh.1  2 2.1  small   low  unacc
0  vhigh   vhigh  2   2  small   med  unacc
1  vhigh   vhigh  2   2  small  high  unacc
2  vhigh   vhigh  2   2    med   low  unacc
3  vhigh   vhigh  2   2    med   med  unacc
4  vhigh   vhigh  2   2    med  high  unacc
5  vhigh   vhigh  2   2    big   low  unacc
6  vhigh   vhigh  2   2    big   med  unacc
7  vhigh   vhigh  2   2    big  high  unacc
8  vhigh   vhigh  2   4  small   low  unacc
9  vhigh   vhigh  2   4  small   med  unacc

Attribute Information->

Class Values: unacc, acc, good, vgood

Attributes->
buying- vhigh, high, med, low.
maint- vhigh, high, med, low.
doors -2, 3, 4, 5more.
persons-2, 4, more.
lug_boot- small, med, big.
safety- low, med, high.

In [4]:
df.describe()

vhigh vhigh.1     2   2.1 small   low  unacc
count   1727    1727  1727  1727  1727  1727   1727
unique     4       4     4     3     3     3      4
top      med     med     3  more   med   med  unacc
freq     432     432   432   576   576   576   1209

In [5]:
# view dimensions of dataset
df.shape

(1727, 7)

We can see that there are 1727 instances and 7 variables in the data set.



**Rename column names**
We can see that the dataset does not have proper column names. The columns are merely labelled as 0,1,2.... and so on. We should give proper names to the columns. I will do it as follows:-

In [6]:
col_names = ['buying', 'maint', 'doors', 'persons', 'lug_boot', 'safety', 'class']
df.columns = col_names
col_names

['buying', 'maint', 'doors', 'persons', 'lug_boot', 'safety', 'class']

In [7]:
df.head()

buying  maint doors persons lug_boot safety  class
0  vhigh  vhigh     2       2    small    med  unacc
1  vhigh  vhigh     2       2    small   high  unacc
2  vhigh  vhigh     2       2      med    low  unacc
3  vhigh  vhigh     2       2      med    med  unacc
4  vhigh  vhigh     2       2      med   high  unacc

Frequency distribution of values in variables
Now, I will check the frequency counts of categorical variable

In [8]:
col_names = ['buying', 'maint', 'doors', 'persons', 'lug_boot', 'safety', 'class']
for col in col_names:
    print(df[col].value_counts())

med      432
low      432
high     432
vhigh    431
Name: buying, dtype: int64
med      432
low      432
high     432
vhigh    431
Name: maint, dtype: int64
3        432
4        432
5more    432
2        431
Name: doors, dtype: int64
more    576
4       576
2       575
Name: persons, dtype: int64
med      576
big      576
small    575
Name: lug_boot, dtype: int64
med     576
high    576
low     575
Name: safety, dtype: int64
unacc    1209
acc       384
good       69
vgood      65
Name: class, dtype: int64


We can see that the doors and persons are categorical in nature. So, I will treat them as categorical variables.

Summary of variables
There are 7 variables in the dataset. All the variables are of categorical data type.
These are given by buying, maint, doors, persons, lug_boot, safety and class.
class is the target variable.

In [12]:
# check missing values in variables
for col in col_names:
    if df[col].isnull().sum()>0:
        print("Missing values in ",col,"are",df[col].isnull().sum())
    else:
        print("No missing values")

No missing values
No missing values
No missing values
No missing values
No missing values
No missing values
No missing values


**Declare feature vector and target variable**

In [13]:
X = df.drop(['class'], axis=1)
y = df['class']

**Split data into separate training and test set**

In [14]:
#  split X and y into training and testing sets

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.33, random_state = 42)

In [15]:
# check the shape of X_train and X_test
print(X_train.shape)
print(X_test.shape)

(1157, 6)
(570, 6)


Feature Engineering

Feature Engineering is the process of transforming raw data into useful features that help us to understand our model better and increase its predictive power. I will carry out feature engineering on different types of variables.

In [16]:
# check data types in X_train
X_train.dtypes

buying      object
maint       object
doors       object
persons     object
lug_boot    object
safety      object
dtype: object

Encode categorical variables
Now, I will encode the categorical variables.

In [17]:
X_train.head()

buying  maint  doors persons lug_boot safety
83    vhigh  vhigh  5more       2      med    low
48    vhigh  vhigh      3    more      med    med
468    high  vhigh      3       4    small    med
155   vhigh   high      3    more      med    low
1043    med   high      4    more    small    low

We can see that all the variables are ordinal categorical data type.



In [18]:
X_train = X_train.apply(LabelEncoder().fit_transform)
X_test = X_test.apply(LabelEncoder().fit_transform)

In [19]:
X_train.head()

buying  maint  doors  persons  lug_boot  safety
83         3      3      3        0         1       1
48         3      3      1        2         1       2
468        0      3      1        1         2       2
155        3      0      1        2         1       1
1043       2      0      2        2         2       1

Now let's perform Machine Learning Classification. We'll be predicting the genre based on the given features.

In [21]:
pip install xgboost

Note: you may need to restart the kernel to use updated packages.


In [22]:
from sklearn.naive_bayes import GaussianNB
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier, XGBRFClassifier
from xgboost import plot_tree, plot_importance
from sklearn.ensemble import AdaBoostClassifier

from sklearn.metrics import confusion_matrix, accuracy_score, roc_auc_score, roc_curve
from sklearn import preprocessing
from sklearn.feature_selection import RFE

Creating a Predefined function to assess the accuracy of a model. This will be the scoring function

In [51]:
def score(model, title = "Default"):
    model.fit(X_train, y_train)
    preds = model.predict(X_test)
    print(confusion_matrix(y_test, preds))
    accuracy = round(accuracy_score(y_test, preds), 5)
    print('Accuracy for', title, ':', accuracy, '\n')

 Let's see how a basic Logistic Regressor works on this!

In [24]:
lg = LogisticRegression(random_state=0, solver='lbfgs', multi_class='multinomial')
score(lg, "Logistic Regression")

Accuracy for Logistic Regression : 0.69825 



**Train the following:**

4. **AdaBoostClassifier** with n_estimators=1000 & random_state=0
5. **RandomForestClassifier** with n_estimators=1000, max_depth=10, random_state=0

**IMPORTANT**: Use the default values for other hyper parameters apart from these. Do **NOT** change these values 

In [25]:
# AdaBoostClassifier
ab = AdaBoostClassifier(n_estimators=1000, random_state=0)
score(ab,"Ada Boost")

Accuracy for Ada Boost : 0.77193 



In [27]:
# Random Forest
rf = RandomForestClassifier(n_estimators=1000,max_depth=10,random_state=0)
score(rf,"Random Forest")

Accuracy for Random Forest : 0.95439 



AdaBoost performs poorly, but Random Forest looks great!

Now look at the import statements. There may be a few models you may or may not have seen. Use those models to beat the score you achieved in the cell above! Play around with their hyper parameters.

In [29]:
# Naive Bayes: GaussianNB
nb = GaussianNB()
score(nb,"Naive Bayes")

# KNN: KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=15)
score(knn,"KNN")

# Decission trees: DecisionTreeClassifier
dt = DecisionTreeClassifier(criterion="gini", max_depth=10, random_state=0)
score(dt,"Decision Tree")

Accuracy for Naive Bayes : 0.61404 

Accuracy for KNN : 0.80877 

Accuracy for Decision Tree : 0.95965 



Decision Tree Classifier with criterion entripy index,max_depth=8, random_state=0

In [36]:
#YOUR CODE HERE
dt = DecisionTreeClassifier(criterion="entropy", max_depth=8, random_state=0)
score(dt,"Decision Tree")

Accuracy for Decision Tree : 0.92632 



Use XGBClassifier with n_estimators=1000 and learning_rate=0.01

In [38]:
# XGBClassifier
xg = XGBClassifier(n_estimators = 1000,learning_rate=0.01 )
score(xg,"XGBClassifier")

[20:19:07] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.5.1/src/learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'multi:softprob' was changed from 'merror' to 'mlogloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
Accuracy for XGBClassifier : 0.97895 



Plot the Confusion Matrix for XGBClassifier with parameters n_estimators=1000, learning_rate=0.1

In [50]:
from sklearn.metrics import classification_report, confusion_matrix

In [58]:
# Cell to plot Confusion Matrix
xg2 = XGBClassifier(n_estimators = 1000,learning_rate=0.01 )
score(xg2,"XGBClassifier")

[20:38:42] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.5.1/src/learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'multi:softprob' was changed from 'merror' to 'mlogloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[[121   4   1   1]
 [  2  13   0   3]
 [  0   0 399   0]
 [  1   0   0  25]]
Accuracy for XGBClassifier : 0.97895 



8. Find the feature importance for XGBClassifier. 

*Hint: It's an inbuilt member variable*

In [59]:
# Cell to find Feature Importance
#YOUR CODE HERE
xg2.feature_importances_

array([0.12472091, 0.14183709, 0.04213865, 0.28286693, 0.08277319,
       0.32566318], dtype=float32)

**Classification Report**
Classification report is another way to evaluate the classification model performance. It displays the precision, recall, f1 and support scores for the model. I have described these terms in later.

We can print a classification for XGBClassifierwith parameters n_estimators=1000, learning_rate=0.1

In [60]:
#YOUR CODE HERE
xg2.fit(X_train, y_train)
preds = xg2.predict(X_test)
print(classification_report(y_test, preds))


[20:39:35] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.5.1/src/learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'multi:softprob' was changed from 'merror' to 'mlogloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
              precision    recall  f1-score   support

         acc       0.98      0.95      0.96       127
        good       0.76      0.72      0.74        18
       unacc       1.00      1.00      1.00       399
       vgood       0.86      0.96      0.91        26

    accuracy                           0.98       570
   macro avg       0.90      0.91      0.90       570
weighted avg       0.98      0.98      0.98       570

